trino-aws-proxy/src/main/java/io/trino/aws/proxy/server/credentials/CredentialsProviderModule.java

```diff
@@ -0,0 +1,81 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.aws.proxy.server.credentials;
+
+import com.google.inject.Binder;
+import com.google.inject.BindingAnnotation;
+import com.google.inject.ConfigurationException;
+import com.google.inject.Inject;
+import com.google.inject.Injector;
+import com.google.inject.Key;
+import com.google.inject.Module;
+import com.google.inject.Provider;
+import com.google.inject.Scopes;
+import com.google.inject.name.Names;
+import io.airlift.configuration.AbstractConfigurationAwareModule;
+import io.trino.aws.proxy.spi.credentials.CredentialsProvider;
+import org.checkerframework.framework.qual.FromByteCode;
+
+import java.lang.annotation.Retention;
+import java.lang.annotation.Target;
+
+import static io.trino.aws.proxy.spi.TrinoAwsProxyBinder.trinoAwsProxyBinder;
+import static java.lang.annotation.ElementType.FIELD;
+import static java.lang.annotation.ElementType.METHOD;
+import static java.lang.annotation.ElementType.PARAMETER;
+import static java.lang.annotation.RetentionPolicy.RUNTIME;
+
+public class CredentialsProviderModule
+       extends AbstractConfigurationAwareModule
+{
+    @Retention(RUNTIME)
+    @Target({FIELD, PARAMETER, METHOD})
+    @BindingAnnotation
+    public @interface ForCredentialsProviderType {
+    }
+
+    @Override
+    protected void setup(Binder binder)
+    {
+        String credentialProviderType = buildConfigObject(CredentialsProviderConfig.class).getCredentialsProviderType();
+        binder.bind(String.class).annotatedWith(ForCredentialsProviderType.class).toInstance(credentialProviderType);
+        trinoAwsProxyBinder(binder)
+                .bindCredentialsProvider(innerBinder -> innerBinder.toProvider(CredentialsProviderBinder.class));
+    }
+
+    public static final class CredentialsProviderBinder
+        implements Provider<CredentialsProvider>
+    {
+        private final Injector injector;
+        private final String credentialsProviderType;
+
+        @Inject
+        public CredentialsProviderBinder(Injector injector, @ForCredentialsProviderType String credentialsProviderType)
+        {
+            this.injector = injector;
+            this.credentialsProviderType = credentialsProviderType;
+        }
+
+        @Override
+        public CredentialsProvider get()
+        {
+            try {
+                return injector.getInstance(Key.get(CredentialsProvider.class, Names.named(credentialsProviderType)));
```

Comment:
I'm -1 on this. For me, this is manually doing what our `TrinoAwsProxyServerPlugin` does using the JDK ServiceLoader. Maybe we can have a Zoom call to discuss.

Reply:
+1 on the Zoom call